### PR TITLE
[SPARK-22490][DOC] Add PySpark doc for SparkSession.builder

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

In PySpark API Document, [SparkSession.build](http://spark.apache.org/docs/2.2.0/api/python/pyspark.sql.html) is not documented and shows default value description.
```
SparkSession.builder = <pyspark.sql.session.Builder object ...
```

This PR adds the doc.

screen

The following is the diff of the generated result.

```
$ diff old.html new.html
95a96,101
> <dl class="attribute">
> <dt id="pyspark.sql.SparkSession.builder">
> <code class="descname">builder</code><a class="headerlink" href="#pyspark.sql.SparkSession.builder" title="Permalink to this definition">¶</a></dt>
> <dd><p>A class attribute having a <a class="reference internal" href="#pyspark.sql.SparkSession.Builder" title="pyspark.sql.SparkSession.Builder"><code class="xref py py-class docutils literal"><span class="pre">Builder</span></code></a> to construct <a class="reference internal" href="#pyspark.sql.SparkSession" title="pyspark.sql.SparkSession"><code class="xref py py-class docutils literal"><span class="pre">SparkSession</span></code></a> instances</p>
> </dd></dl>
>
212,216d217
< <dt id="pyspark.sql.SparkSession.builder">
< <code class="descname">builder</code><em class="property"> = &lt;pyspark.sql.session.SparkSession.Builder object&gt;</em><a class="headerlink" href="#pyspark.sql.SparkSession.builder" title="Permalink to this definition">¶</a></dt>
< <dd></dd></dl>
<
< <dl class="attribute">
```

## How was this patch tested?

Manual.

```
cd python/docs
make html
open _build/html/pyspark.sql.html
```

### CODE REVIEW

File: python/docs/pyspark.sql.rst
Comment (dongjoon-hyun) [inline python/docs/pyspark.sql.rst]: This blocks the default dummy doc for `builder`

File: python/pyspark/sql/session.py
Comment (dongjoon-hyun) [inline python/pyspark/sql/session.py]: This adds a correct doc.

**Comment by dongjoon-hyun** [comment]
cc @gatorsmile

File: python/docs/pyspark.sql.rst
Comment (HyukjinKwon) [reply python/docs/pyspark.sql.rst]: I believe this is a bug in Sphinx. I think it should automatically detect the attribute docstring. I recently took a look before:

- `https://github.com/sphinx-doc/sphinx/pull/2299`
- `https://github.com/sphinx-doc/sphinx/issues/2298`

I haven't tested the fix by myself but I believe this will be fixed in Sphinx 1.6.6.

File: python/pyspark/sql/session.py
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]: I think we should rather resemble:

https://github.com/apache/spark/blob/d6ee69e7761a62e47e21e4c2ce1bb20038d745b6/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L970

and I'd use the term "class attribute" instead of "class variable"`

**Review by HyukjinKwon — COMMENTED** [review]
I am okay with this if this is an exceptional case for attribute docstring.

But I think we should generally avoid the workaround with `exclude-members` and manual `autoattribute` because If I understood correctly, it is fixed in the coming Sphinx 1.6.6.

Please let me know if I missed anything here.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @HyukjinKwon .

File: python/pyspark/sql/session.py
Comment (dongjoon-hyun) [reply python/pyspark/sql/session.py]: Sure.

Comment (dongjoon-hyun) [reply python/pyspark/sql/session.py]: Yep.

**Comment by dongjoon-hyun** [comment]
@HyukjinKwon . For upgrading `sphinx` version, we can upgrade the minimal version from [1.2](https://github.com/apache/spark/blob/master/python/docs/conf.py#L26) to 1.6.6. We had better upgrade in another issue because we may need to consider Jenkins worker environments. Or, is it safe if I do that here?

**Comment by HyukjinKwon** [comment]
Oh, I had to be clear, I think Sphinx 1.6.6 is in development and upgrading it is actually a bigger issue I think.

I tried other ways I could think ahead actually, but failed to figure out a better way. So, I meant LGTM assuming we should generally avoid this workaround but it's one exception (or possibly few in the future), and it is by a bug existing in Sphinx 1.6.6 <.

But let's wait for more feedback just in case I missed something and someone has a better idea.

**Comment by dongjoon-hyun** [comment]
I see. Thank you, @HyukjinKwon .

**Comment by gatorsmile** [comment]
cc @ueshin

**Comment by ueshin** [comment]
LGTM for now, but I'd like you to add some comments to explain this is a workaround for Sphinx <`1.6.6`.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @gatorsmile and @ueshin .

**Review by HyukjinKwon — APPROVED** [review]

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.2

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon, @ueshin , @gatorsmile .

**Pull request closed without merging**
